pkg/resource/deploy/providers/registry.go

```diff
@@ -266,6 +266,13 @@ func (r *Registry) Diff(urn resource.URN, id resource.ID, olds, news resource.Pr
 	label := fmt.Sprintf("%s.Diff(%s,%s)", r.label(), urn, id)
 	logging.V(7).Infof("%s: executing (#olds=%d,#news=%d)", label, len(olds), len(news))
 
+	// If we have no old state, use the old inputs as the old state. This handles the scenario where the CLI is being
+	// upgraded from a version that did not reflect provider inputs to provider outputs, and a provider is being
+	// upgraded from a version that did not implement DiffConfig to a version that does.
+	if len(olds) == 0 {
+		olds = news
+	}
```

Comment:
slightly concerned about the semantics of "no old" versus "empty old".  Is it the case that all resources have at least *one* property?  If so, this is ok. Otherwise, say you had a resource with no props, and you moved to a version where props were added.  It feels like we would then say "we should use the `news`, instead of sticking with the empty `olds`.

Thoughts?

Reply:
> Is it the case that all resources have at least one property? If so, this is ok. Otherwise, say you had a resource with no props, and you moved to a version where props were added. It feels like we would then say "we should use the news, instead of sticking with the empty olds.

We recently began reflecting the input properties as the output properties for all provider resources. As a result, the only way a provider resource can have no output properties in the checkpoint is if it had no input properties when it was created, or if it was created prior to the changes that began reflecting inputs to outputs.

In the former case, both old inputs and old outputs will be empty, so there will be no semantic difference with these changes.

In the latter case, the old inputs will not be empty, but the old state will be empty. This can cause unexpected diffs when dealing with providers that implement `DiffConfig` (e.g. the k8s provider): because these providers are presented an empty old state, they are very likely to see a difference between the empty old state and the current inputs, even if the user's program has not changed. Reflecting the old inputs into the new state is logically performing an upgrade from the older format to the newer format.

All that said, I realize that the change I made above isn't quite right--I need to move the upgrade code around a bit.

Reply:
Fixed.